marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochSPD.hs

```diff
@@ -0,0 +1,442 @@
+      :: forall proto era c
+       . (c ~ Ledger.Crypto era, c ~ O.StandardCrypto)
+      => O.LedgerState (O.ShelleyBlock proto era)
+      -> Map C.PoolId C.Lovelace
+    getStakeMapFromShelleyBlock st = spd
+      where
+        nes = O.shelleyLedgerState st :: Ledger.NewEpochState era
+
+        stakeSnapshot = Ledger._pstakeSet . Ledger.esSnapshots . Ledger.nesEs $ nes :: Ledger.SnapShot c
+
+        stakes = Ledger.unStake
+               $ Ledger._stake stakeSnapshot
+
+        delegations :: VMap.VMap VMap.VB VMap.VB (Ledger.Credential 'Ledger.Staking c) (Ledger.KeyHash 'Ledger.StakePool c)
+        delegations = Ledger._delegations stakeSnapshot
+
+        spd :: Map C.PoolId C.Lovelace
+        spd = Map.fromListWith (+)
+            $ map swap
+            $ catMaybes
+            $ VMap.elems
+            $ VMap.mapWithKey
+                (\cred spkHash ->
+                    (\c -> ( C.Lovelace $ coerce $ Ledger.fromCompact c
+                           , C.StakePoolKeyHash spkHash
+                           )
+                    )
+                    <$> VMap.lookup cred stakes)
+                delegations
+
+getEpochNo :: O.LedgerState (O.CardanoBlock O.StandardCrypto) -> Maybe EpochNo
+getEpochNo ledgerState' = case ledgerState' of
+  O.LedgerStateByron _st  -> Nothing
+  O.LedgerStateShelley st -> getEpochNoFromShelleyBlock st
+  O.LedgerStateAllegra st -> getEpochNoFromShelleyBlock st
+  O.LedgerStateMary st    -> getEpochNoFromShelleyBlock st
+  O.LedgerStateAlonzo st  -> getEpochNoFromShelleyBlock st
+  O.LedgerStateBabbage st -> getEpochNoFromShelleyBlock st
+  where
+    getEpochNoFromShelleyBlock = Just . Ledger.nesEL . O.shelleyLedgerState
+
+data EpochSPDRow = EpochSPDRow
+    { epochSPDRowEpochNo         :: !C.EpochNo
+    , epochSPDRowPoolId          :: !C.PoolId
+    , epochSPDRowLovelace        :: !C.Lovelace
+    , epochSPDRowSlotNo          :: !C.SlotNo
+    , epochSPDRowBlockHeaderHash :: !(C.Hash C.BlockHeader)
+    , epochSPDRowBlockNo         :: !C.BlockNo
+    } deriving (Eq, Show, Generic, SQL.FromRow, SQL.ToRow)
+
+-- We should only store on disk SPD from the last slot of each epoch.
+instance Buffered EpochSPDHandle where
+    persistToStorage
+        :: Foldable f
+        => f (StorableEvent EpochSPDHandle)
+        -> EpochSPDHandle
+        -> IO EpochSPDHandle
+    persistToStorage events h@(EpochSPDHandle c ledgerStateDirPath securityParam) = do
+        let eventsList = toList events
+        -- Only keep the event with the largest slot number of each epoch.
+        let eventsToStore =
+                mapMaybe (\es -> listToMaybe
+                               $ List.sortOn (\e -> Down $ epochSPDEventSlotNo e) es)
+                  $ List.groupBy (\e1 e2 -> epochSPDEventEpochNo e1 == epochSPDEventEpochNo e2)
+                  eventsList
+
+        forM_ (concatMap eventToEpochSPDRows eventsToStore) $ \row ->
+          SQL.execute c
+              [r|INSERT OR REPLACE INTO epoch_spd
```

Comment:
the Replace works when there is a unique constraint on the table, as a [constraint violation resolution](https://www.sqlite.org/lang_conflict.html).  
I don't think this is what we want here